client/state/experiments/reducer.ts

```diff
@@ -0,0 +1,68 @@
+/**
+ * External Dependencies
+ */
+import { Action } from 'redux';
+
+/**
+ * Internal Dependencies
+ */
+import { EXPERIMENT_FETCH, EXPERIMENT_ASSIGN, CURRENT_USER_RECEIVE } from 'state/action-types';
+import { ExperimentState, ExperimentAssign } from 'state/experiments/types';
+import { tracksAnonymousUserId } from 'lib/analytics/ad-tracking';
+
+/**
+ * Attempt to get the anon id for the user, if set
+ */
+export function getAnonIdFromCookie(): string | null {
+	const id = tracksAnonymousUserId();
+
+	return id == null || id === '' ? null : id;
+}
+
+const appStartedAt = Date.now();
+
+const resetState: ( anonId: string | null ) => ExperimentState = anonId => ( {
+	anonId,
+	isLoading: true,
+	nextRefresh: appStartedAt,
+	variations: null,
+} );
+
+export default function reducer( state: ExperimentState = resetState( null ), action: Action ) {
+	switch ( action.type ) {
+		/**
+		 * Store the anon id when we first load the application
+		 */
+		case '@@INIT':
+			state.anonId = getAnonIdFromCookie();
+			break;
+
+		/**
+		 * Store the user's assignment from the API
+		 */
+		case EXPERIMENT_ASSIGN:
+			state = {
+				...state,
+				isLoading: false,
+				variations: ( action as ExperimentAssign ).variations,
+				nextRefresh: ( action as ExperimentAssign ).nextRefresh,
```

Comment:
It should be possible to define the `ExperimentAssign` type in such a way that the compiler infers the type automatically from the fact that we are in a `case` where `action.type === EXPERIMENT_ASSIGN`. @sirreal?